### PR TITLE
feat(query): Add jq set returning function for processing Variant data with jq filters

### PR DESCRIPTION
**Pull request by maxjustus**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

This adds a `jq` set returning function for working with Variant data. Jq is described as "regexp for json" and is a wonderful addition for processing data in Variant columns. This work was inspired by [pgjq](https://tselai.com/pgjq-dsl-database.html) which adds jq to postgres and [litejq](https://github.com/Florents-Tselai/liteJQ) which adds jq to sqlite.

the `jq` function being set returning means that you can write jq filters that unnest json data as a part of their evaluation. For example:

```
SELECT * FROM jq('.[] as $a | {upcased: ($a | ascii_upcase), original: $a}', json_array('hello', 'world'))
```
```
row 1: {"upcased": "HELLO", "original": "hello"}
row 2: {"upcased": "WORLD", "original": "world"}
```

It uses the lightweight [jaq](https://lib.rs/crates/jaq) crate which provides a high performance pure rust implementation of jq.

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16288)

### CODE REVIEW

**Comment by maxjustus** [comment]
Rebased against most recent main, fixed clippy and cargo machete errors.

**Comment by xudong963** [comment]
> Rebased against most recent main, fixed clippy and cargo machete errors.

Thanks for your contribution.  You can run `REGENERATE_GOLDENFILES=1 cargo test` to update the failed unit test.

File: src/query/functions/src/srfs/variant.rs [hunk 1]
```diff
@@ -460,6 +470,162 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+                            .map(|e| format!("err: {} location: {:?}", e.0, e.1))
+                            .collect::<Vec<_>>()
+                            .join("\n");
+                        ctx.set_error(0, err_str);
+                        return vec![];
+                    }
+
+                    let jaq_args = vec![];
+                    // You can pass additional scalar inputs as args to the jq filter.
+                    // This could be a useful enhancement, but leaving it out for mow.
```
Comment (b41sh) [inline src/query/functions/src/srfs/variant.rs:530]: `for mow` seems a typo.

File: src/query/functions/src/srfs/variant.rs [hunk 2]
```diff
@@ -460,6 +470,162 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+    });
+}
+
+// This comes straight from the jaq source. It converts a Jaq val type to a JSON string.
+// https://github.com/01mf02/jaq/blob/426fdab46c95e7ed0dadc5c049b3d83388271b1a/jaq/src/main.rs#L511
+// there may be an opportunity to format directly into
+// jsonb data so it doesn't need to be re-parsed.
+fn jaq_val_to_json(val: &Val) -> String {
+    match val {
+        Val::Null => "null".to_string(),
```
Comment (b41sh) [inline src/query/functions/src/srfs/variant.rs:599]: We can convert `Val` to `jsonb::Value` and then convert to binary format.
for `Array` and `Object`, we can use function `jsonb::build_array` and `jsonb::build_object`.
[cast_scalar_to_variant](https://github.com/datafuselabs/databend/blob/main/src/query/expression/src/types/variant.rs#L198) function is an example to look at.

**Review by b41sh — APPROVED** [review]
LGTM, @maxjustus
Thanks for your contribution, this feature is very useful.

**Comment by BohuTANG** [comment]
Thanks @maxjustus, for your contribution!

We should add this new function to the docs. @soyeric128 @b41sh

**Pull request merged**